discovery/manager.go

```diff
@@ -191,6 +191,7 @@ func (m *Manager) ApplyConfig(cfg map[string]sd_config.ServiceDiscoveryConfig) e
 		}
 	}
 	m.cancelDiscoverers()
+	failedConfigs.WithLabelValues(m.name).Set(0)
```

Comment:
You'll also need to change the line that currently increments the counter when a failed config is encountered. Instead, it should always be set to the current number of failed SD configs, not a cumulative count over time.

Reply:
Not quite sure about your question. Can you check my test within the PR, to see if that's your requirement ?

And there was the call to `Inc()`, I just keep it.
(https://github.com/prometheus/prometheus/blob/e4a741cb7dfa24978097971c2c8d5c8cc12e6cef/discovery/manager.go#L335)

Reply:
Ah, ok. I was confused because it's uncommon to use `Inc()` instead of `Set()` on a gauge, but in this case it looks like the right thing because you're setting it to `0` every time the config is applied, and `m.registerProviders()` is only called from `ApplyConfig()`, so it should be fine.

Reply:
That seems a bit racy.

Reply:
Indeed, the problem is that a scrape could happen while the Prometheus server is iterating over configs. It will then underreport the number of failed configs. On the other hand, we are doing something similar with `discoveredTargets` below.

It would be better to let `registerProviders` return the number of failed configs and set them once after the loop in L203.

Reply:
I guess you mean ? @beorn7 
```go
failedCount := 0

for name, scfg := range cfg {
	failedCount = m.registerProviders(scfg, name)
	discoveredTargets.WithLabelValues(m.name, name).Set(0)
}

failedConfigs.WithLabelValues(m.name).Set(failedCount)
```

Reply:
That might work. But you want `failedCount += m.registerProviders(scfg, name)`, don't you?

Reply:
Yeah, exactly `+=` 😸

Reply:
Due to the signature change of `registerProviders` (will add a return value), I've checked all the references of this method, only used in one place in this file.🤞

Reply:
Sounds all good to me. @cstyan , would be great if you could sanity-check this once @nevill has updated the PR.

And many thanks @nevill for following up.